core/src/main/scala/kafka/log/Log.scala

```diff
@@ -609,24 +716,22 @@ class Log(val dir: File,
    * logSegment will be rolled if one of the following conditions met
    * <ol>
    * <li> The logSegment is full
-   * <li> The maxTime has elapsed
+   * <li> The maxTime has elapsed since the largest timestamp in the segment (or since the create time if there is no
+   * message having a timestamp)
    * <li> The index is full
    * </ol>
    * @return The currently active segment after (perhaps) rolling to a new segment
    */
   private def maybeRoll(messagesSize: Int): LogSegment = {
     val segment = activeSegment
+    val reachedRollMs = segment.timeWaitedForRoll(time.milliseconds) > config.segmentMs - segment.rollJitterMs
```

Comment:
I am not sure if we have the same guarantee even if we still let the log rolling based on the log segment create time. There are a few issues,

1. When the timestamp type is CreateTime, the messages in the segment may have older timestamp than the create time of the log segment. So the log rolling may still be delayed from the application perspective.
2. In some linux system, the create time of a file is not available. So based on segment file create time may not always work.

I am thinking that maybe we can just let the log rolling based on the timestamp of the first time index entry. We can always insert an index entry when the first message is appended to the log segment. This provides similar guarantee of the previous log rolling and does not have the above two issues. What do you think?